src/librustc/ty/sty.rs

```diff
@@ -927,11 +888,27 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
         }
     }
 
-    pub fn is_empty(&self, _cx: TyCtxt) -> bool {
-        // FIXME(#24885): be smarter here
+    pub fn is_never(&self) -> bool {
+        match self.sty {
+            TyNever => true,
+            _ => false,
+        }
+    }
+
+    pub fn is_uninhabited(&self, _cx: TyCtxt) -> bool {
+        // FIXME(#24885): be smarter here, the AdtDefData::is_empty method could easily be made
+        // more complete.
         match self.sty {
             TyEnum(def, _) | TyStruct(def, _) => def.is_empty(),
-            _ => false
+            TyNever => true,
+
+            // FIXME(canndrew): There's no reason why this can't be uncommented, it's tested and it
+            // doesn't break anything. But I'm keeping my changes small for now.
+            //TyTuple(ref tys) => tys.iter().any(|ty| ty.is_uninhabited(cx)),
+
+            // FIXME(canndrew): this line breaks core::fmt
+            //TyRef(_, ref tm) => tm.ty.is_uninhabited(cx),
+            _ => false,
```

Comment:
IMO this function should remain untouched, it *specifically* looks for `enum Foo {}` and *any* change in it could have unintended consequences that we are *not* prepared for.